src/tqec/generation/circuit.py

```diff
@@ -70,20 +67,107 @@ def generate_circuit(
+
+
+def _create_mapping(
+    plaquette: Plaquette,
+    scheduled_circuit: ScheduledCircuit,
+    offset: Position,
+) -> dict[cirq.Qid, cirq.Qid]:
+    origin = plaquette.origin
+    assert origin == Position(0, 0), "Only origin (0,0) is supported for now"
+
+    qubit_map = {
+        # GridQubit are indexed as (row, col), so (y, x)
+        # Qubits are given relative to an origin, so we need to add the offset
+        qubit: qubit + (offset.y, offset.x) + (origin.y, origin.x)  # type: ignore
+        for qubit in scheduled_circuit.raw_circuit.all_qubits()
+    }
+    return qubit_map
+
+
+def _find_offset(
+    scheduled_circuits: list[ScheduledCircuit],
+    row_index: int,
+    column_index: int,
+    length: int,
+    default_x_increment: int,
+    default_y_increment: int,
+) -> Position:
+    """Computes the offset by looking at already placed circuits
+
+    We anchor the top left corner by one of the following ways:
+        - at the origin (0,0) (first circuit)
+        - the top right corner of the previous circuit (left)
+        - the bottom left corner of the previous circuit (above)
+        - the bottom ove the above and the right of the left (above and left)
+    """
+    # first circuit
+    if len(scheduled_circuits) == 0 or row_index == column_index == 0:
+        return Position(0, 0)
+    default_multiplier = 0
+    for circuit in scheduled_circuits[: -(column_index + 1) : -1]:
+        if len(circuit.raw_circuit.all_qubits()) == 0:
+            default_multiplier += 1
+        else:
+            position_left = _find_coordinate(circuit, above=False)
+            position_left.x += default_x_increment * default_multiplier
+            break
+    else:
+        position_left = Position(default_x_increment * default_multiplier, 0)
+
+    if row_index == 0:  # left
+        return Position(position_left.x, 0)
+
+    default_multiplier = 0
+    for circuit in reversed(scheduled_circuits[column_index::length]):
+        if len(circuit.raw_circuit.all_qubits()) == 0:
+            default_multiplier += 1
+        else:
+            position_above = _find_coordinate(circuit, above=True)
+            position_above.y += default_y_increment * default_multiplier
+            break
+    else:
+        position_above = Position(0, default_y_increment * default_multiplier)
+
+    if column_index == 0:  # above
+        return position_above
+    if position_left == position_above:
+        return position_left
+    return Position(position_left.x, position_above.y)
+
+
+def _find_coordinate(final_circuit: ScheduledCircuit, above: bool = False) -> Position:
+    if len(final_circuit.raw_circuit.all_qubits()) == 0:
+        return Position(0, 0)
+    position = max(
+        (
+            Position(qubit.col, qubit.row)
+            for qubit in final_circuit.raw_circuit.all_qubits()
+        ),
+        key=lambda p: (p.y, -p.x) if above else (p.x, -p.y),
+    )
+    return Position(position.x, position.y)
```

Comment:
I have hard times understanding the usefulness of these functions. In my mind, with the pre-conditions we imposed, the global position of any plaquette origin should be 
```
(column_index * default_x_increment, row_index * default_y_increment)
```
with potentially a small offset to avoid to top-most and left-most plaquettes to have qubits with negative coordinates if we want to avoid that. 
Am I missing something?

Reply:
The functions push a circuit to the top left of its two neighboring circuits that have already been placed.
From my understanding, using
```
(column_index * default_x_increment, row_index * default_y_increment)
```
implicitly assumes equally shaped, but rectangular plaquettes; am I wrong?
Say we have a plaquette of shape `(6,3)` and one of `(3,3)`, depending who is placed first, `defualt_x_increment` would be either 2 or 4?

Reply:
The formula above only assumes that the distance between 2 neighbouring plaquette origins is `default_x_increment` in the `x` dimension and `default_y_increment` in the `y` dimension. The actual plaquette shapes are not part of this formula, and 2 neighbouring plaquettes can even overlap if the user implemented them that way. 

Also, `default_{x,y}_increment` should be constant for the whole template. 

I am also having hard times thinking about examples of non-constant plaquettes. I will ask an example to @afowler, it might help me understand what should be done better.

Reply:
I wasn't ware they could overlap, I thought they could only have touching borders, but if this is the case, then your solution is definitely a lot simpler.
I was thinking of general tilings where something like [this](https://www.pngfind.com/pngs/m/70-705325_truncated-square-tiling-hexagon-and-square-tile-hd.png) would be possible.

Reply:
Let's wait for Austin's feedback on this. If I remember correctly, he said that overlapping could be a thing and that it was up to the user to check that the circuits provided are correct.
The tiling you gave as an example could be implemented with `default_x_increment == default_y_increment` no?

Reply:
IMO `default_x_increment`/`default_y_increment` should be some properties specified by the templates and the plaquettes not aware of. All the mission of the templates should be providing a valid region of qubits and a pattern that the plaquettes can be placed on. Whether the plaquettes can be overlapping(BTW I think this can happen) should do nothing to the coordinating process.  The interop between the plaquettes on the templates only happens when validating the scheduled circuits, where the operation should be unique on every qubits at each moment.